From d0ff412380cc78d5175eac5b6f612aa1e4dead87 Mon Sep 17 00:00:00 2001
From: Anton Bambura <jenneron@protonmail.com>
Date: Fri, 22 Jan 2021 22:40:16 +0200
Subject: [PATCH 172/210] arm64: dts: qcom: msm8916-wingtech-wt865x8: add sound
 and modem (v4)

This patch adds sound support for these devices.
This patch adds reserved-memory ranges and additional modem trees for
these devices.

Signed-off-by: Anton Bambura <jenneron@protonmail.com>

MODEM:
v2: Setup memshare / add GPS memory
v3: wcnss/venus no longer need to be relocated

v4: Combine sound and modem commits and drop extra -modem.dts
    The no-modem variant is now patched dynamically by lk2nd instead.

Signed-off-by: Stephan Gerhold <stephan@gerhold.net>
---
 .../dts/qcom/msm8916-wingtech-wt86518.dts     | 46 +++++++++++++++
 .../dts/qcom/msm8916-wingtech-wt86528.dts     | 59 +++++++++++++++++++
 .../dts/qcom/msm8916-wingtech-wt865x8.dtsi    | 39 ++++++++++++
 3 files changed, 144 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86518.dts b/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86518.dts
index 9237cc168d8..5ad69e88cf0 100644
--- a/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86518.dts
+++ b/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86518.dts
@@ -8,6 +8,17 @@ / {
 	model = "Lenovo A6000 (Wingtech WT86518)";
 	compatible = "wingtech,wt86518", "qcom,msm8916";
 	chassis-type = "handset";
+
+	speaker_amp: audio-amplifier {
+		compatible = "awinic,aw8738";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&spk_ext_pa_default>;
+
+		enable-gpios = <&msmgpio 119 GPIO_ACTIVE_HIGH>;
+		aw8738,mode = <1>;
+		sound-name-prefix = "Speaker Amp";
+	};
 };
 
 &blsp_i2c2 {
@@ -24,6 +35,18 @@ accelerometer@e {
 	};
 };
 
+&gps_mem {
+	status = "disabled";
+};
+
+&headphones_switch {
+	VCC-supply = <&pm8916_l17>;
+};
+
+&memshare {
+	status = "disabled";
+};
+
 &pm8916_charger {
 	status = "okay";
 
@@ -47,6 +70,19 @@ &reg_lcd_pos {
 	regulator-max-microvolt = <5400000>;
 };
 
+&sound {
+	status = "okay";
+
+	model = "wt88047";
+	audio-routing =
+		"Speaker Amp IN", "HPH_R_EXT",
+		"Headphones Switch INL", "HPH_L",
+		"Headphones Switch INR", "HPH_R",
+		"AMIC1", "MIC BIAS Internal1",
+		"AMIC2", "MIC BIAS Internal2";
+	aux-devs = <&speaker_amp>, <&headphones_switch>;
+};
+
 &usb {
 	dr_mode = "peripheral";
 	extcon = <&pm8916_charger>;
@@ -55,3 +91,13 @@ &usb {
 &usb_hs_phy {
 	extcon = <&pm8916_charger>;
 };
+
+&msmgpio {
+	spk_ext_pa_default: spk-ext-pa-default {
+		pins = "gpio119";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-disable;
+	};
+};
diff --git a/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86528.dts b/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86528.dts
index 7cc043db75d..5ddd31d39cb 100644
--- a/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86528.dts
+++ b/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt86528.dts
@@ -9,6 +9,30 @@ / {
 	compatible = "wingtech,wt86528", "qcom,msm8916";
 	chassis-type = "handset";
 
+	/* left AW8736 */
+	speaker_amp_left: audio-amplifier-left {
+		compatible = "awinic,aw8738";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&spk_ext_pa_left_default>;
+
+		enable-gpios = <&msmgpio 119 GPIO_ACTIVE_HIGH>;
+		aw8738,mode = <3>;
+		sound-name-prefix = "Speaker Amp L";
+	};
+
+	/* right AW8736 */
+	speaker_amp_right: audio-amplifier-right {
+		compatible = "awinic,aw8738";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&spk_ext_pa_right_default>;
+
+		enable-gpios = <&msmgpio 121 GPIO_ACTIVE_HIGH>;
+		aw8738,mode = <3>;
+		sound-name-prefix = "Speaker Amp R";
+	};
+
 	gpio-leds {
 		compatible = "gpio-leds";
 
@@ -76,6 +100,21 @@ &reg_lcd_pos {
 	regulator-max-microvolt = <5000000>;
 };
 
+&sound {
+	status = "okay";
+
+	model = "wt86528";
+	audio-routing =
+		"Speaker Amp L IN", "HPH_L_EXT",
+		"Speaker Amp R IN", "HPH_R_EXT",
+		"Headphones Switch INL", "HPH_L",
+		"Headphones Switch INR", "HPH_R",
+		"AMIC1", "MIC BIAS External1",
+		"AMIC2", "MIC BIAS Internal2",
+		"AMIC3", "MIC BIAS External1";
+	aux-devs = <&speaker_amp_left>, <&speaker_amp_right>, <&headphones_switch>;
+};
+
 &usb {
 	extcon = <&usb_id>, <&usb_id>;
 };
@@ -84,6 +123,10 @@ &usb_hs_phy {
 	extcon = <&usb_id>;
 };
 
+&wcd_codec {
+	qcom,micbias1-ext-cap;
+};
+
 &msmgpio {
 	gpio_leds_default: gpio-leds-default {
 		pins = "gpio16", "gpio17";
@@ -101,6 +144,22 @@ imu_default: imu-default {
 		bias-disable;
 	};
 
+	spk_ext_pa_left_default: spk-ext-pa-left-default {
+		pins = "gpio119";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-disable;
+	};
+
+	spk_ext_pa_right_default: spk-ext-pa-right-default {
+		pins = "gpio121";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-disable;
+	};
+
 	usb_id_default: usb-id-default {
 		pins = "gpio110";
 		function = "gpio";
diff --git a/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt865x8.dtsi b/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt865x8.dtsi
index d2659627d56..02cf1334b1f 100644
--- a/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt865x8.dtsi
+++ b/arch/arm64/boot/dts/qcom/msm8916-wingtech-wt865x8.dtsi
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: GPL-2.0-only
 
 #include "msm8916-pm8916.dtsi"
+#include "msm8916-modem.dtsi"
+
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/pinctrl/qcom,pmic-mpp.h>
 
@@ -13,6 +15,28 @@ chosen {
 		stdout-path = "serial0";
 	};
 
+	reserved-memory {
+		mpss_mem: mpss@86800000 {
+			reg = <0x0 0x86800000 0x0 0x5500000>;
+			no-map;
+		};
+
+		gps_mem: gps@8bd00000 {
+			reg = <0x0 0x8bd00000 0x0 0x200000>;
+			no-map;
+		};
+	};
+
+	headphones_switch: audio-switch {
+		compatible = "simple-audio-amplifier";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&headphones_switch_default>;
+
+		enable-gpios = <&msmgpio 120 GPIO_ACTIVE_HIGH>;
+		sound-name-prefix = "Headphones Switch";
+	};
+
 	backlight: backlight {
 		compatible = "pwm-backlight";
 		pwms = <&pm8916_pwm 0 100000>;
@@ -190,6 +214,13 @@ &usb {
 	status = "okay";
 };
 
+&wcd_codec {
+	qcom,micbias-lvl = <2800>;
+	qcom,mbhc-vthreshold-low = <75 150 237 450 500>;
+	qcom,mbhc-vthreshold-high = <75 150 237 450 500>;
+	qcom,hphl-jack-type-normally-open;
+};
+
 &smd_rpm_regulators {
 	vdd_l1_l2_l3-supply = <&pm8916_s3>;
 	vdd_l4_l5_l6-supply = <&pm8916_s4>;
@@ -302,6 +333,14 @@ gpio_keys_default: gpio-keys-default {
 		bias-pull-up;
 	};
 
+	headphones_switch_default: headphones-switch-default {
+		pins = "gpio120";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-disable;
+	};
+
 	mdss {
 		mdss_default: mdss-default {
 			pins = "gpio25";
-- 
2.34.1

